soc: nordic: common: `CONFIG_SOC_NRF_FORCE_CONSTLAT`

JordanYates opened the pull request:
Move the option to force constant latency mode outside of nRF54l, since it is an option applicable to most Nordic SoCs.

This is a quality-of-life option to prevent boards/applications from needing to implement the calls themselves.

JordanYates committed:
soc: nordic: common: `CONFIG_SOC_NRF_FORCE_CONSTLAT`

Move the option to force constant latency mode outside of nRF54l, since
it is an option applicable to most Nordic SoCs.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates committed:
samples: nordic: nrf_sys_event: test `SOC_NRF_FORCE_CONSTLAT`

Validate that `CONFIG_SOC_NRF_FORCE_CONSTLAT=y` compiles.

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented on soc/nordic/common/nrf_constlat.c:
```diff
@@ -0,0 +1,17 @@
+}
+
+/* Immediately after the SoC init functions */
+SYS_INIT(nrf_const_lat, PRE_KERNEL_1, 1);
```
I had a look around for an appropriate place and couldn't find one. For `soc.c` it would need to duplicated for each Nordic SoC. If that is preferred I can do that.

cvinayak replied:
We need to use on-off service and port Bluetooth Controllers to use them. Otherwise, low power mode would get set after radio events.

JordanYates committed:
bluetooth: controller: ll_sw: nordic: constant latency req

When `CONFIG_NRF_SYS_EVENT` is enabled, route constant latency requests
through the reference counted API.

Signed-off-by: Jordan Yates <jordan@embeint.com>

nordic-krch commented on soc/nordic/common/Kconfig:
```diff
@@ -21,6 +21,13 @@ config NRF_SYS_EVENT
 	bool "nRF system event support"
 	select NRFX_POWER if !NRF_PLATFORM_HALTIUM
 
+config NRF_FORCE_CONSTANT_LATENCY
```
There is already a feature like that in nRF54L familiy.
https://github.com/zephyrproject-rtos/zephyr/blob/9ae10ea9b216c609a31e0398c479ae1a701616af/soc/nordic/nrf54l/Kconfig#L79

Can you rename it to SOC_NRF_FORCE_CONSTLAT here and remove the Kconfig from nrf54l/Kconfig.

JordanYates replied:
I did not know that existed, and based on comments from @cvinayak the option is probably broken when used with a Bluetooth controller

nordic-krch reviewed (commented):
Dupilicate code from nrf54l/soc.c would need to be removed: https://github.com/zephyrproject-rtos/zephyr/blob/9ae10ea9b216c609a31e0398c479ae1a701616af/soc/nordic/nrf54l/soc.c#L142

bjarki-andreasen commented on soc/nordic/common/Kconfig:
```diff
@@ -21,6 +21,13 @@ config NRF_SYS_EVENT
 
+config NRF_FORCE_CONSTANT_LATENCY
+	bool "Force constant latency mode in system ON"
+	depends on NRF_SYS_EVENT
```
`NRF_SYS_EVENT` is available to the NRF54H20 app and radio cores, so only !RISCV would be helpful.

nordicjm replied:
`!RISCV` ? I have this being changed in the register directly from the flpr core and it doesn't seem to cause problems so that condition seems wrong:
```
	/* Switch to constant latency */
//NRF_POWER_NS_BASE
	NRF_POWER->TASKS_LOWPWR = 0;
	NRF_POWER->TASKS_CONSTLAT = 1;
k_sleep(K_SECONDS(2));
```

bjarki-andreasen replied:
If sys event is supported on RISCV cores, we should probably extend SYS_EVENT?

JordanYates commented:
Updated PR to pull the already existing option out to the common level, so current users on nRF54L won't see any changes.
